system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -63,70 +101,422 @@ class CLibCompiledXpath : public CInterface, public ICompiledXpath
     {
         return m_compiledXpathExpression;
     }
+
+    virtual void extractReferences(StringArray &functions, StringArray &variables) override
+    {
+        char *buf = nullptr;
+        size_t len = 0;
+
+        FILE *stream = open_memstream(&buf, &len);
+        if (stream == nullptr)
+            return;
+
+        xmlXPathDebugDumpCompExpr(stream, m_compiledXpathExpression, 0);
+        fputc(0, stream);
+        fflush(stream);
+        fclose (stream);
+        char *line = buf;
+        while (line)
+            line = extractFromLineGetNext(functions, variables, line);
+        free (buf);
+    }
+};
+static xmlXPathObjectPtr variableLookupFunc(void *data, const xmlChar *name, const xmlChar *ns_uri);
+
+typedef std::map<std::string, xmlXPathObjectPtr> XPathObjectMap;
+
+class CLibXpathScope
+{
+public:
+    StringAttr name; //in future allow named parent access?
+    XPathObjectMap variables;
+
+public:
+    CLibXpathScope(const char *_name) : name(_name){}
+    ~CLibXpathScope()
+    {
+        for (XPathObjectMap::iterator it=variables.begin(); it!=variables.end(); ++it)
+            xmlXPathFreeObject(it->second);
+    }
+    bool setObject(const char *key, xmlXPathObjectPtr obj)
+    {
+        std::pair<XPathObjectMap::iterator,bool> ret = variables.emplace(key, obj);
+        if (ret.second==true)
+            return true;
+        //within scope, behave exactly like xmlXPathContext variables are added now, which seems to be that they are replaced
+        //if we're preventing replacing variables we need to handle elsewhere
+        //  and still replace external values when treated as xsdl:variables, but not when treated as xsdl:params
+        if (ret.first->second)
+            xmlXPathFreeObject(ret.first->second);
+        ret.first->second = obj;
+        return true;
+    }
+    xmlXPathObjectPtr getObject(const char *key)
+    {
+        XPathObjectMap::iterator it = variables.find(key);
+        if (it == variables.end())
+            return nullptr;
+        return it->second;
+    }
 };
 
-class CLibXpathContext : public CInterface, public IXpathContext
+typedef std::vector<std::unique_ptr<CLibXpathScope>> XPathScopeVector;
+typedef std::map<std::string, ICompiledXpath*> XPathInputMap;
+
+class XpathContextState
 {
 private:
+    xmlDocPtr doc = nullptr;
+    xmlNodePtr node = nullptr;
+    int contextSize = 0;
+    int proximityPosition = 0;
+public:
+    XpathContextState(xmlXPathContextPtr ctx)
+    {
+        doc = ctx->doc;
+        node = ctx->node;
+        contextSize = ctx->contextSize;
+        proximityPosition = ctx->proximityPosition;
```

Comment:
Those values tell the context where you are in terms of processing the current node set.  If you look at my test case when I use the "position()=1" function it works because these were set properly.
contextSize is the number of nodes in the set.
proximityPosition is what node in the set we are working on.

```
<es:for-each select="Alias">
     <es:choose>
         <es:when test="position()=1">
```

Reply:
So when we move back up the stack we need to restore the current working positions to the previous level.

Reply:
between this and the variable scopes it is much more efficient then copying the entire context.. and it's similar to the way the library is used in libxslt.